### PR TITLE
Add Muon Optimizer Support to ZeRO3

### PR DESCRIPTION
**Pull request by PKUWZP**
Kudos to @pengdurice for the great work and discussions! 

We aim on adding Muon Optimizer to zero stage 3 in this PR: 

- Created a dedicated momentum buffer in zero stage 3 optimizer to save the momentum buffers specifically for Muon Optimizer.  
- The optimizer states can be dispatched into 3 devices: GPU, CPU and NVME. For GPU and CPU, we just make the new buffers the same device of `self.fp32_partitioned_groups_flat`; when `device == NVME`, we make sure that the momentum buffers can be swapped in and out along with other components in the optimizer states. 
- The new momentum buffers are also partitioned like `self.fp32_partitioned_groups_flat` to save memory footprint. So, before the muon update, we need to perform `all_gather` on top of each data-parallel  group rank. The Muon updates of the parameters are also divided across the data-parallel ranks, and the results are all-gathered once all updates are complete. After the `all_gather`, the momentum buffers are partitioned and flatted again.

Next steps:
- Explore quantization of momentum buffers for saving memory
- Explore using highly optimized Adam / AdamW Optimizers

### CODE REVIEW

File: deepspeed/runtime/zero/stage3.py [hunk 1]
```diff
@@ -1420,6 +1456,119 @@ def __reduce_and_partition_ipg_grads(self, communication_data_type: torch.dtype)
...
+            params_size_offset += param.grad.numel()
+        # if optimizer is swappable, swap in the momentum buffer of the parameters that need to be updated using muon and then swap them out
+        # if optimizer is not swappable, find the momentum buffer of the parameters that need to be updated using muon in memory
+        for i in params_to_subgroup_maps:
+            if self._swappable_optimizer_subgroup(i) and not self.save_muon_momentum_buffer_in_memory:
+                self.optimizer_swapper.swap_in_optimizer_state(parameter=self.fp32_partitioned_groups_flat[i])
+                for idx, dest_offset in params_to_subgroup_maps[i]:
+                    momentum_buffer[idx] = self.optimizer.state[
+                        self.fp32_partitioned_groups_flat[i]]["momentum_buffer"].narrow(
+                            0, dest_offset, param.partition_numel()).clone()
```
Comment (PKUWZP) [inline deepspeed/runtime/zero/stage3.py:1499]: @pengdurice Here is a bug. The variable param refers to the last parameter from the previous loop (`for param in self.ipg_buckets[...]`), not the parameter corresponding to idx. We should change it to`use_muon_params[idx].partition_numel()`.

File: deepspeed/runtime/zero/stage3.py [hunk 2]
```diff
@@ -330,11 +331,16 @@ def _enforce_optimizer_offload():
...
+        self.save_muon_momentum_buffer_in_memory = ds_config.get('save_muon_momentum_buffer_in_memory', False)
+        if self.use_muon and self.reduce_scatter:
+            raise ValueError("Muon Optimizer and reduce scatter are not compatible.")
+        if self.use_muon and self.all2all_process_group is not None:
+            raise ValueError("Muon and all2all process group are not compatible.")
+
         self.dp_process_group = self.parameter_offload.dp_process_group
         self.sequence_parallel_size = groups._get_sequence_parallel_world_size()
 
-        self.all2all_process_group = all2all_process_group
```
Comment (PKUWZP) [inline deepspeed/runtime/zero/stage3.py:336]: @pengdurice Question: where did we set up the `all2all_process_group`? It seems that it's never set.

File: deepspeed/runtime/zero/stage3.py [hunk 3]
```diff
@@ -1420,6 +1456,119 @@ def __reduce_and_partition_ipg_grads(self, communication_data_type: torch.dtype)
...
+                world_sz = dist.get_world_size(self.dp_process_group)
+                rank = dist.get_rank(self.dp_process_group)
+                params = [use_muon_params[idx] for idx, _ in params_to_subgroup_maps[i]]
+                gathered_momentums = [gathered_params_momentums[idx] for idx, _ in params_to_subgroup_maps[i]]
+                # params_pad = params + [torch.empty_like(params[-1])] * (world_sz - len(params) % world_sz)
+                grads_pad = [param.grad for param in params
+                             ] + [torch.empty_like(params[-1].grad)] * (world_sz - len(params) % world_sz)
+                gathered_momentums_pad = gathered_momentums + [torch.empty_like(gathered_momentums[-1])
+                                                               ] * (world_sz - len(gathered_momentums) % world_sz)
+                for base_i in range(len(params))[::world_sz]:
```
Comment (PKUWZP) [inline deepspeed/runtime/zero/stage3.py:1530]: @pengdurice There's a padding error here. When `len(params) % world_sz == 0`, this adds `world_sz` empty tensors instead of 0. Should we change it to: `(world_sz - len(params) % world_sz) % world_sz` ?

File: deepspeed/runtime/zero/stage3.py [hunk 4]
```diff
@@ -330,11 +331,16 @@ def _enforce_optimizer_offload():
 
         self.reduce_scatter = reduce_scatter
 
+        self.use_muon = 'muon' in self.optimizer.__class__.__name__.lower()
```
Comment (PKUWZP) [inline deepspeed/runtime/zero/stage3.py:334]: @pengdurice This is very fragile and purely depends on the naming conventions. Can we leverage `isinstance()` instead?

File: deepspeed/runtime/zero/stage3.py [hunk 5]
```diff
@@ -330,11 +331,16 @@ def _enforce_optimizer_offload():
...
+        self.save_muon_momentum_buffer_in_memory = ds_config.get('save_muon_momentum_buffer_in_memory', False)
```
Comment (PKUWZP) [inline deepspeed/runtime/zero/stage3.py:335]: @pengdurice Can we add `save_muon_momentum_buffer_in_memory` to the config schema and documented?

File: deepspeed/runtime/zero/stage3.py [hunk 6]
```diff
@@ -1420,6 +1456,119 @@ def __reduce_and_partition_ipg_grads(self, communication_data_type: torch.dtype)
...
+                param.grad.data.copy_(buffer_to_reduce.narrow(0, params_size_offset,
+                                                              param.grad.numel()).view_as(param.grad),
+                                      non_blocking=False)
+                momentum_buffer.append(None)
+                if i not in params_to_subgroup_maps:
+                    params_to_subgroup_maps[i] = []
+                params_to_subgroup_maps[i].append((idx, dest_offset))
+                idx += 1
+            params_size_offset += param.grad.numel()
+        # if optimizer is swappable, swap in the momentum buffer of the parameters that need to be updated using muon and then swap them out
```
Comment (PKUWZP) [inline deepspeed/runtime/zero/stage3.py:1491]: @pengdurice This doubles NVMe I/O overhead. Can we consider consolidating into a single swap in/out cycle?

File: deepspeed/runtime/zero/stage3.py [hunk 7]
```diff
@@ -1420,6 +1456,119 @@ def __reduce_and_partition_ipg_grads(self, communication_data_type: torch.dtype)
...
+                )
+        # if there are parameters that need to be updated using muon
+        if momentum_buffer:
+            # all gather the momentum buffers of the parameters to the global buffer
+            # this is done since the momentum buffers are stored in partitions just like the params themselves
+            gathered_params_momentums = self._partitioned_buffers_all_gather(use_muon_params, momentum_buffer,
+                                                                             communication_data_type)
+            for i in params_to_subgroup_maps:
+                if self._swappable_optimizer_subgroup(i) and not self.save_muon_momentum_buffer_in_memory:
+                    self.optimizer_swapper.swap_in_optimizer_state(parameter=self.fp32_partitioned_groups_flat[i])
```
Comment (PKUWZP) [inline deepspeed/runtime/zero/stage3.py:1517]: @pengdurice Again same thing here, can we consolidate the two swaps into one swap?

**Review by PKUWZP — COMMENTED** [review]
@pengdurice Also two more comments: 

- It seems that we have excessive tensor allocations: Multiple torch.empty, torch.zeros, and .clone() calls create memory footprint pressure. Consider reusing buffers where possible.

- Synchronous `all_gather`: The distributed operations could potentially be overlapped with computation.

I think we need to re-work on the PR and let's take some times to refine the code.
